src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+  // best rendition by player dimension
+  while (i--) {
+    variant = bandwidthPlaylists[i];
+
+    // ignore playlists without resolution information
+    if (!variant.attributes ||
+        !variant.attributes.RESOLUTION ||
+        !variant.attributes.RESOLUTION.width ||
+        !variant.attributes.RESOLUTION.height) {
+      continue;
+    }
+
+    // since the playlists are sorted, the first variant that has
+    // dimensions less than or equal to the player size is the best
+
+    let variantResolution = variant.attributes.RESOLUTION;
+
+    if (variantResolution.width === width &&
+        variantResolution.height === height) {
+      // if we have the exact resolution as the player use it
+      resolutionPlusOne = null;
+      resolutionBestVariant = variant;
+      break;
+    } else if (variantResolution.width < width &&
+               variantResolution.height < height) {
+      // if both dimensions are less than the player use the
+      // previous (next-largest) variant
+      break;
+    } else if (!resolutionPlusOne ||
+               (variantResolution.width < resolutionPlusOneAttribute.width &&
+                variantResolution.height < resolutionPlusOneAttribute.height)) {
+      // If we still haven't found a good match keep a
+      // reference to the previous variant for the next loop
+      // iteration
+
+      // By only saving variants if they are smaller than the
+      // previously saved variant, we ensure that we also pick
+      // the highest bandwidth variant that is just-larger-than
+      // the video player
+      resolutionPlusOne = variant;
+      resolutionPlusOneAttribute = resolutionPlusOne.attributes.RESOLUTION;
+    }
+  }
+
+  // fallback chain of variants
+  return resolutionPlusOne ||
+    resolutionBestVariant ||
+    bandwidthBestVariant ||
+    sortedPlaylists[0];
+};
+
+export default class MasterPlaylistController extends Stream {
+  constructor({url, withCredentials, currentTimeFunc, mediaSourceMode, hlsHandler,
+    externHls}) {
+    super();
+
+    Hls = externHls;
+
+    this.withCredentials = withCredentials;
+    this.currentTimeFunc = currentTimeFunc;
+    this.mediaSourceMode = mediaSourceMode;
+
+    this.mediaSource = new videojs.MediaSource({ mode: this.mediaSourceMode });
+    // load the media source into the player
+    this.mediaSource.addEventListener('sourceopen', this.handleSourceOpen_.bind(this));
+
+    // combined audio/video or just video when alternate audio track is selected
+    this.mainSegmentLoader_ = new SegmentLoader({
+      mediaSource: this.mediaSource,
+      currentTime: this.currentTimeFunc,
+      withCredentials: this.withCredentials
+    });
+    // alternate audio track
+    this.audioSegmentLoader_ = new SegmentLoader({
+      mediaSource: this.mediaSource,
+      currentTime: this.currentTimeFunc,
+      withCredentials: this.withCredentials
+    });
+
+    this.hlsHandler = hlsHandler;
```

Comment:
The main reason we are passing hlsHandler right now is to allow users to override what they previously could override. For instance, the selectPlaylist function. At the moment people can set that to be whatever function they want, and expect it to be called with the HLS Handler. We could pass the bound function down instead, which prevents MasterPlaylistController from having the entire hlsHandler reference, but it might not buy us too much in terms of separation, and if the user updates it again, we'd have to call back in to MPC and modify it (which is fine, we'd just have to agree on that interaction). Were you thinking of a different relationship?